rrtypes.go

```diff
@@ -0,0 +1,374 @@
+	//
+	// then the client would connect to 192.0.2.2:1111. But if you had the
+	// following URL:
+	//
+	//  https://example.com:2222/
+	//
+	// then the client would lookup the following records:
+	//
+	//  _2222._https.example.com.  60  IN  HTTPS  1  example.net.  alpn=h2,h3
+	//  example.net.               60  IN  A      192.0.2.3
+	//
+	// and the client would connect to 192.0.2.3:2222. And if you had the same
+	// URL but the following records:
+	//
+	//  _2222._https.example.com.  60  IN  HTTPS  1  example.net.  alpn=h2,h3 port=3333
+	//  example.net.               60  IN  A      192.0.2.4
+	//
+	// then the client would connect to 192.0.2.4:3333.
+	//
+	// So the key things to note here are that:
+	//
+	//  - If you want to change the port that the client connects to, you need
+	//    to set the “port=” value in the “Params” field, not the
+	//    “URLSchemePort”.
+	//
+	//  - The client will never lookup the HTTPS record prefixed with the
+	//    underscored default port, so you should only set “URLSchemePort” if
+	//    you are explicitly using a non-default port in the URL.
+	//
+	//  - It is completely valid to set the “port=” value in the “Params” field
+	//    to the default port for the scheme, but also completely unnecessary.
+	//
+	//  - The “URLSchemePort” field and the “port=” value in the “Params” field
+	//    are completely independent, with one exception: if you set the
+	//    “URLSchemePort” field to a non-default port and leave the “port=”
+	//    value in the “Params” field unset, then the client will default to the
+	//    value of the “URLSchemePort” field, and not to the default port for
+	//    the scheme.
+	URLSchemePort uint16
+
+	Name string
+	TTL  time.Duration
+
+	// “Priority” is the priority of the service, with lower values indicating
+	// that clients should prefer this service over others.
+	//
+	// Note that Priority==0 is a special case, and indicates that the record
+	// is an “Alias” record. Alias records behave like CNAME records, but are
+	// allowed at the root of a zone. When in Alias mode, the Params field
+	// should be unset.
+	Priority uint16
+
+	// “Target” is the target of the service, which is typically a hostname or
+	// an alias (CNAME or other SVCB record). If this field is set to a single
+	// dot ".", then the target is the same as the name of the record (without
+	// the underscore-prefixed components, of course).
+	Target string
+
+	// “Params” is a map of key–value pairs that are used to specify various
+	// parameters for the service. The keys are typically registered with IANA,
+	// and which keys are valid is service-dependent.
+	//
+	// Note that there is a key called “mandatory”, but this does not mean that
+	// it is mandatory for you to set the listed keys. Instead, this means that
+	// if a client does not understand all of the listed keys, then it must
+	// ignore the entire record. This is similar to the “critical” flag in CAA
+	// records.
+	Params *SvcParams
```

Comment:
```suggestion
	Params SvcParams
```

As explained, probably best to keep this a value not a pointer. (Even though it makes comparison difficult.)

Reply:
> > So maybe it might be better to instead define
> > `Value    *SvcParams`
> > That should stop it from panicking, but I don't know enough about Go to know the greater implications to making a struct field a pointer.
>
> Hmm, that would indeed do a comparison, but it'll do a pointer comparison, nothing actually useful. It will only compare if the memory addresses are the same, not if the semantic value of the pointer is the same. So that's only useful if you want to ensure you're using the same instance/copy of the SvcParams. It doesn't check semantic meaning.

Right, but I think that giving a false comparison is probably (maybe?) better than a runtime panic. Using the following test file:

```go
package main

import (
	"context"
	"fmt"
	"os"

	"github.com/libdns/libdns"
	"github.com/libdns/rfc2136"
)

func doStuffWithRecord(rec libdns.Record) {
	fmt.Printf("Found record: %#v\n", rec)
}

func main() {
	ctx := context.Background()
	provider := rfc2136.Provider{
		KeyName: os.Getenv("RFC2136_KEYNAME"),
		KeyAlg:  os.Getenv("RFC2136_KEYALG"),
		Key:     os.Getenv("RFC2136_KEY"),
		Server:  os.Getenv("RFC2136_SERVER"),
	}
	zone := os.Getenv("RFC2136_ZONE")

	fmt.Println("Getting records from provider...")
	recs, err := provider.GetRecords(ctx, zone)
	if err != nil {
		panic(err)
	}

	fmt.Printf("Records: %#v\n", recs)

	for _, targetRecord := range recs {
		targetRecord, _ := targetRecord.RR().Parse()
		for _, rec := range recs {
			rec, _ := rec.RR().Parse()
			if rec == targetRecord {
				doStuffWithRecord(rec)
			}
		}
	}
	fmt.Println("End of records")
}
```

with `Params *SvcParams` I get

```log
Getting records from provider...
Records: []libdns.Record{libdns.RR{Name:"@", TTL:0, Type:"SOA", Data:"ns.maxchernoff.ca. dns.maxchernoff.ca. 2025040404 0 0 0 0"}, libdns.TXT{Name:"@", TTL:3600000000000, Text:"just a TXT record"}, libdns.SVCB{Scheme:"https", URLSchemePort:0x0, Name:"@", TTL:3600000000000, Priority:0x1, Target:".", Params:(*libdns.SvcParams)(0xc000124080)}, libdns.RR{Name:"@", TTL:0, Type:"SOA", Data:"ns.maxchernoff.ca. dns.maxchernoff.ca. 2025040404 0 0 0 0"}}
Found record: libdns.RR{Name:"@", TTL:0, Type:"SOA", Data:"ns.maxchernoff.ca. dns.maxchernoff.ca. 2025040404 0 0 0 0"}
Found record: libdns.RR{Name:"@", TTL:0, Type:"SOA", Data:"ns.maxchernoff.ca. dns.maxchernoff.ca. 2025040404 0 0 0 0"}
Found record: libdns.TXT{Name:"@", TTL:3600000000000, Text:"just a TXT record"}
Found record: libdns.RR{Name:"@", TTL:0, Type:"SOA", Data:"ns.maxchernoff.ca. dns.maxchernoff.ca. 2025040404 0 0 0 0"}
Found record: libdns.RR{Name:"@", TTL:0, Type:"SOA", Data:"ns.maxchernoff.ca. dns.maxchernoff.ca. 2025040404 0 0 0 0"}
End of records
```

but with `Params SvcParams` I get

```log
Getting records from provider...
Records: []libdns.Record{libdns.RR{Name:"@", TTL:0, Type:"SOA", Data:"ns.maxchernoff.ca. dns.maxchernoff.ca. 2025040404 0 0 0 0"}, libdns.TXT{Name:"@", TTL:3600000000000, Text:"just a TXT record"}, libdns.SVCB{Scheme:"https", URLSchemePort:0x0, Name:"@", TTL:3600000000000, Priority:0x1, Target:".", Params:libdns.SvcParams{"alpn":[]string{"h2", "h3"}}}, libdns.RR{Name:"@", TTL:0, Type:"SOA", Data:"ns.maxchernoff.ca. dns.maxchernoff.ca. 2025040404 0 0 0 0"}}
Found record: libdns.RR{Name:"@", TTL:0, Type:"SOA", Data:"ns.maxchernoff.ca. dns.maxchernoff.ca. 2025040404 0 0 0 0"}
Found record: libdns.RR{Name:"@", TTL:0, Type:"SOA", Data:"ns.maxchernoff.ca. dns.maxchernoff.ca. 2025040404 0 0 0 0"}
Found record: libdns.TXT{Name:"@", TTL:3600000000000, Text:"just a TXT record"}
panic: runtime error: comparing uncomparable type libdns.SVCB

goroutine 1 [running]:
main.main()
        /home/max/Projects/libdns-rfc2136/_nogit-testing/main.go:38 +0x4c5
exit status 2
```

Now, This is admittedly a contrived example, so maybe this isn't relevant in regular code?

Reply:
Good point, that code _looks_ like it should work. (But `RR().Parse()` is a bit of a weird pattern, yeah?)

Thinking on this more, it seems like a bad idea to try to allow primitive comparison of `Record` values.

Extending my argument about `SvcParams` not being obviously comparable more broadly, is it reasonable to suggest that equality of an entire, arbitrary `Record` (whatever type it is) isn't defined/obvious?

`RR` yes, but arbitrary `Record` values, no, I don't believe so...

So, maybe this is a good thing to bring up in the docs: `Record` values shouldn't be used for comparison. It's not obvious what equality means in terms of a DNS record, and it may be different for everyone.

If people really wanted to compare records primitively, I think they can compare the RR structs. Otherwise I think it's best that they write their own comparison method.

Reply:
> But `RR().Parse()` is a bit of a weird pattern, yeah?

Hmm, now that you mention it, that might actually be necessary in a lot of cases right now. It would be valid for an implementation to do something like return `libdns.Address` structs for `A`/`AAAA` RRs and `libdns.Record` for all other record types, so if you want to make sure that you always have a specific record type (for a type switch), you'd have to run `RR().Parse()` first. 

Should we maybe specify that `libdns` implementations should always return the most specific record type possible?

> Extending my argument about `SvcParams` not being obviously comparable more broadly,

In the DNS (binary) wire format, the `SvcParams` keys are required to be increasing numeric order of their numeric value ([RFC 9460 §2.2](https://datatracker.ietf.org/doc/html/rfc9460#section-2.2-5), so there is a canonical/official way to compare equality. But if we wanted to implement this, we'd need to hard-code in the [list of valid keys](https://www.iana.org/assignments/dns-svcb/dns-svcb.xhtml) to make sure that we sort them properly.

> It's not obvious what equality means in terms of a DNS record, and it may be different for everyone.

Because DNSSEC signs the hash of a record, each RR type has an official/canonical representation, so I'd say that that would be the “obvious” definition of equality, but it would be a lot of work for us to implement those semantics.

> If people really wanted to compare records primitively, I think they can compare the RR structs. Otherwise I think it's best that they write their own comparison method.

:+1:, I think that this is the best solution.

> Thinking on this more, it seems like a bad idea to try to allow primitive comparison of `Record` values.
>
> [...]
>
> So, maybe this is a good thing to bring up in the docs: `Record` values shouldn't be used for comparison.

Ok, I'll add this to the documentation. What should we do with the `Params` field though? I would still lean towards `Params *SvcParams` because otherwise, if `libdns` implementations ignore the documentation and primitively compare `Record` values, we could end up in a place where everything works fine until someone adds a `HTTPS` RR to the DNS and then their program starts panicking. 

For a worst-case scenario, someone could enable ECH in Caddy, which then adds an `HTTPS` RR, which then makes the `libdns` implementation panic, which then crashes Caddy. But then even if they disabled ECH after the first crash, the `HTTPS` RR would still be there, so Caddy would keep crashing.

But that being said, the comparison with `Params *SvcParams` would be bogus, so maybe it would make more sense to use `Params SvcParams` so that the error would be obvious?

Reply:
> Hmm, now that you mention it, that might actually be necessary in a lot of cases right now. It would be valid for an implementation to do something like return libdns.Address structs for A/AAAA RRs and libdns.Record for all other record types, so if you want to make sure that you always have a specific record type (for a type switch), you'd have to run RR().Parse() first.

Ahh, good point.

So I had a big long reply drafted up to this, documenting my thought process that you spurred on, but then my browser tab closed and I lost it.

Basically... yeah, let's document that providers should return specific type structs rather than RR.

There's just no way to technically enforce and require that they do without being evil:

```go
// remove RR() from the RR struct, then:

type Record interface {
    RR() RR
    mustUseOurTypes() unexportedType
}
```

This is the only way I can think of that would prevent them from defining their own types that implement the Record interface (making type assertions difficult again), and force that they don't return `RR` structs.

(I also like the flexibility (for input purposes) of RR also being a Record.)

Reply:
> I see what you're saying, but, pointers to maps are quite unconventional. I also think it's generally unwise to primitively compare interface types. At least the panic alerts you to incorrect code, but using a pointer will just hide the bug. I think it should not be a pointer.

Ok, that totally works for me. I think that I agree with you now that the panic is actually good since it lets you know about the error right away instead of hiding it somewhere in the code. Thanks for hearing me out though.